### PR TITLE
Add unique userID support to authentication framework

### PR DESCRIPTION
**Pull request by madurangasiriwardena**

### CODE REVIEW

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java [hunk 1]
```diff
@@ -209,6 +224,7 @@ public String getUserId(String userName, int tenantId, String userDomain) throws
      * @return the list of user Ids of users stored in the given user store
      * @throws UserSessionException if an error occurs when retrieving the user id list from the database
      */
+    @Deprecated
```
Comment (senthalan) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java:227]: Better if we can add  why this is deprecated and what should be the alternative method to use (if any)

File: components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/cache/BaseCacheTest.java [hunk 2]
```diff
@@ -31,6 +31,7 @@
 import java.io.File;
 import java.net.URL;
 
+import static org.mockito.Matchers.any;
```
Comment (omindu) [inline components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/cache/BaseCacheTest.java:34]: Do we need this change ?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/GraphBasedSequenceHandlerExceptionRetryTest.java [hunk 3]
```diff
@@ -57,8 +57,11 @@
 
     private static final String CONTEXT_ATTRIBUTE_NAME_CURRENT_FAIL_TRIES = "RetriesOnTest";
 
-    public void testExceptionRetry() throws
-            Exception {
+    public void testExceptionRetry() throws Exception {
+
+        PrivilegedCarbonContext.getThreadLocalCarbonContext()
+                .setTenantDomain(MultitenantConstants.SUPER_TENANT_DOMAIN_NAME);
+        PrivilegedCarbonContext.getThreadLocalCarbonContext().setTenantId(MultitenantConstants.SUPER_TENANT_ID);
```
Comment (omindu) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/GraphBasedSequenceHandlerExceptionRetryTest.java:64]: This is already done in L66+

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/GraphBasedSequenceHandlerClaimsTest.java [hunk 4]
```diff
@@ -70,15 +74,13 @@ public void testHandleClaimHandling() throws Exception {
 
         UserCoreUtil.setDomainInThreadLocal("test_domain");
 
-        RealmService currentRealmService = FrameworkServiceDataHolder.getInstance().getRealmService();
-
         RealmService mockRealmService = mock(RealmService.class);
         UserRealm mockUserRealm = mock(UserRealm.class);
-        UserStoreManager mockUserStoreManager = mock(UserStoreManager.class);
+        AbstractUserStoreManager mockUserStoreManager = PowerMockito.mock(AbstractUserStoreManager.class);
```
Comment (omindu) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/GraphBasedSequenceHandlerClaimsTest.java:79]: Can't we keep the type as UserStoreManager ?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/GraphBasedSequenceHandlerClaimMappingsTest.java [hunk 5]
```diff
@@ -80,23 +82,25 @@ public void testHandleClaimHandling() throws Exception {
 
         RealmService mockRealmService = mock(RealmService.class);
         UserRealm mockUserRealm = mock(UserRealm.class);
-        UserStoreManager mockUserStoreManager = mock(UserStoreManager.class);
+        AbstractUserStoreManager mockUserStoreManager = PowerMockito.mock(AbstractUserStoreManager.class);
```
Comment (omindu) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/GraphBasedSequenceHandlerClaimMappingsTest.java:85]: Can we keep the type as `UserStoreManager` ?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/AuthenticatedUser.java [hunk 6]
```diff
@@ -78,6 +89,29 @@ public AuthenticatedUser(AuthenticatedUser authenticatedUser) {
         }
     }
 
+    public AuthenticatedUser(org.wso2.carbon.user.core.common.User user) {
+        this.userId = user.getUserID();
```
Comment (omindu) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/AuthenticatedUser.java:93]:
```suggestion

        this.userId = user.getUserID();
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java [hunk 7]
```diff
@@ -168,13 +181,15 @@ public String getUserId(String userName, int tenantId, String userDomain, int id
 
     /**
      * Method to return the user Id of a user from the database.
```
Comment (omindu) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java:183]: Can we improve the method comment here ?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java [hunk 8]
```diff
@@ -457,7 +457,7 @@ private static String getServiceProviderMappedUserRoles(ApplicationConfig applic
 
         ClaimManager claimManager = getClaimManager(tenantDomain, realm);
 
-        UserStoreManager userStore = getUserStoreManager(tenantDomain, realm, authenticatedUser.getUserStoreDomain());
+        AbstractUserStoreManager userStore = getUserStoreManager(tenantDomain, realm);
```
Comment (omindu) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java:460]: Please verify whether only userId based methods are invoked through `userStore`

**Review by pulasthi7 — APPROVED** [review]

**Pull request merged**